Adds host parsing

jeremykendall opened the pull request:
Here's the merge of php-domain-parser and Aura.Uri.  The good news is it passes all the test cases, and I _think_ most of it makes good sense.

Here are my immediate areas of concern:
- **Class Names and Locations**: I'm not sure how you guys feel about having the PublicSuffixList\* classes mixed in with the other top level Uri classes.  I'm not sure that makes sense, but I wasn't sure where else to put them so they went there.
- **HttpAdapter Interface and class**: I know that Aura.Http exists, but I also know it's important for each component to be independent. The adapter I've added here is nice and simple, but 1) it's not an Aura.Http adapter and 2) it's lifted from Will Durand's [Geocoder](https://github.com/willdurand/Geocoder) and as such we'd have to give him credit and include his MIT license.  
  - I don't have a problem with crediting Will, but I don't know if you want to have to do that in Aura. We'd have to come up with another solution if not.
  - We'd probably want to add ext-curl as a requirement in composer.json. Does that violate the no external dependency rule?
- ~~**vfsStream dev dependency**: I've added vfsStream as an external dev dependency in composer.json.  This allows for mocking the file system during some tests.  Does that violate the no dependency rule?~~
- ~~**/tests/bootstrap.php**: The tests bootstrap now relies on composer to autoload vfsStream~~
- **Composer vendor binary**: In order to allow end-user updating of the Public Suffix List, I've added a bin directory and have added that bin to the composer.json file.
- **Derivative works**: There are methods in my php-domain-parser that are derived from work in https://github.com/usrflo/registered-domain-libs. Because those are integral to parsing, they're here too.  See Aura\Uri\Host::getRegisterableDomain() and the phpdoc comments for an example.  How should we deal with licensing, attribution, etc in that case?  Or should I refactor the code so that it's not derivative?

I'm sure there are some other areas that we'll need to discuss, and some possible refactoring to do, before finally merging this.  I just wanted to raise the above points because they nagged at me during development.

jeremykendall committed:
Removes leftover todo annotations

jeremykendall committed:
Adds a 'vendor binary' for refreshing cached Public Suffix List

jeremykendall committed:
Adds vfsStream as dev dependency, adds vendor bin

jeremykendall committed:
Adds default Public Suffix List cache folder

* Default cache dir allows for creating new PublicSuffixListManager without
specifying a cache dir
* Providing copies of the Public Suffix List allows users to start using
Aura.Uri immediately, without having to set up cache directories and without
having to explicitly download and cache lists
* The vendor bin `aura-psl` is provided for refreshing the lists

jeremykendall committed:
Adds new classes to src.php

jeremykendall committed:
Updates and adds new classes and tests for parsing hosts

jeremykendall committed:
Refactors test bootstrap

Sets error reporting to -1 and turns on display errors
Defines a test root constant to aid in finding /tests/_files/*
Makes use of Composer autoloader to ease autoloading of vfsStream

jeremykendall committed:
CS fixes via php-cs-fixer

$ php-cs-fixer --verbose fix src/
   1) Aura/Uri/Host.php (trailing_spaces)
   2) Aura/Uri/HttpAdapter/CurlHttpAdapter.php (trailing_spaces)
   3) Aura/Uri/HttpAdapter/HttpAdapterInterface.php (trailing_spaces)
   4) Aura/Uri/Path.php (trailing_spaces, return, controls_spaces)
   5) Aura/Uri/PublicSuffixList.php (trailing_spaces)
   6) Aura/Uri/PublicSuffixListManager.php (trailing_spaces)
   7) Aura/Uri/Query.php (trailing_spaces)
   8) Aura/Uri/Url/Factory.php (trailing_spaces)
   9) Aura/Uri/Url.php (trailing_spaces, return)

harikt commented:
About `ext-curl` , I feel we can keep in suggest. So it is only required for PublicSuffix ? Or may be we can comment on it.

Normally `vfsstream` is added in tests folder itself by @pmjones . If you look into the `Aura.View` or `Aura.Http` you can see https://github.com/auraphp/Aura.View/tree/develop/tests/org .

I also think it is good that we can place `vfsstream` outside for we are not shipping `phpunit` . And as `phpunit` is available by `composer` may be we can relay on `composer` .

Anyway the authority is @pmjones :+1:  . 

About MIT : I feel MIT is compatible with BSD , no ? If so I am not sure whether we want to keep MIT also. I have no trouble. And I don't understand licence better :-1:  .

Thank you for the hard work on this.

jeremykendall commented:
@harikt:

> About ext-curl , I feel we can keep in suggest. So it is only required for PublicSuffix ? Or may be we can comment on it.

I didn't consider that.  And curl is only used by the default Http adapter.

> Normally vfsstream is added in tests folder itself by @pmjones . If you look into the Aura.View or Aura.Http you can see https://github.com/auraphp/Aura.View/tree/develop/tests/org.

Genius! I'll refactor to comply with that precedent.

> About MIT : I feel MIT is compatible with BSD , no?

I'm hoping @pmjones will be able to address the "inclusion of derived works" question both from a licensing/attribution standpoint and from an Aura culture and standards standpoint.

harikt commented:
I have a question : 

Is it a good idea to keep the domain parser as a standalone repo. Wonders what will happen when we want to use the [Aura.Filter](https://github.com/auraphp/Aura.Filter) to validate a domain .

jeremykendall committed:
Removes dev dependency from composer, updates test bootstrap accordingly

Moves vfsStream into tests directory and out of the composer dev requirements per
a conversation with @harikt in #13.

jeremykendall committed:
Removed them from composer, forgot to commit. /facepalm

jeremykendall committed:
Removes all that hand rolled parsing

As it turns out, parse_url will only fail on complex URLs if the scheme is
missing.  Parser now prefixes the url with http:// if no scheme exists.
